bip-txhash.md

```diff
@@ -0,0 +1,269 @@
+  * `TXFS_OUTPUTS_ALL = TXFS_OUTPUTS_SCRIPTPUBKEYS | TXFS_OUTPUTS_VALUES`
+
+For both inputs and then outputs, do the following:
+
+* If the "in/outputs" field is set to 1, another additional byte is expected:
+  * The highest bit (`TXFS_INOUT_NUMBER`) indicates whether the "number of in-/outputs"
+    should be committed to.
+  * For the remaining bits, there are three exceptional values:
+    * 0x00 (`TXFS_INOUT_SELECTION_NONE`) means "no in/outputs"
+      (hence only the number of them as `0x80` (`TXFS_INOUT_NUMBER`)).
+    * `0x40` (`TXFS_INOUT_SELECTION_CURRENT`) means "select only the in/output of the current input index"
+      (it is invalid when current index exceeds number of outputs).
+    * `0x3f` (`TXFS_INOUT_SELECTION_ALL`) means "select all in/outputs".
+
+  * The second highest bit (`TXFS_INOUT_SELECTION_MODE`) is the "specification mode":
+    * Set to 0 it means "leading mode".
+    * Set to 1 it means "individual mode".
+  * In "leading mode", the third highest bit (`TXFS_INOUT_SELECTION_SIZE`) is
+    used to indicate the "count size", i.e. the number of bytes will be used to
+    represent the number of in/output.
+    * With "index size" set to 0, the remaining lowest 5 bits of the first byte will
+      be interpreted as the number of leading in/outputs to select.
+    * With "index size" set to 1, the remaining lowest 5 bits of the first byte together with the
+      8 bits of the next byte will be interpreted as the number of leading in/outputs to select.
+  * In "individual mode", the remaining lowest 6 bits of the first byte will be
+    interpreted as `n`, the number of individual in/outputs to select. For each
+    individual input, (at least) one byte is expected, of this byte. The
+    highest bit is used to indicate "absolute or relative" indices.
+    * If the highest bit is set to 0, it is an absolute index. The second
+      highest bit is used to indicate the amount of bytes are used to represent
+      the index.
+      * If the second-highest bit is 0, the remaining 6 bits represent the index to be selected.
+      * If the second-highest bit is 1, the remaining 6 bits, together with the 8 bits of the next
+        byte, represent the index to be selected.
+    * If the highest bit is set to 1, it is a relative index. The second highest bit is used to
+      indicate the sign of the index.
+      * If the second-highest bit is set to 0, the remaining 6 bits represent the positive relative
+        index to be selected.
+      * If the second-highest bit is set to 1, the remaining 6 bits represent the negative relative
+        index to be selected.
+
+Effectively, this allows a user to select
+* all in/outputs
+* the current input index
+* the leading in/outputs up to 8192
+* up to 64 individually selected in/outputs
+** using absolute indices up to 16384
+** using indices relative to the current input index from -64 to +64.
```

Comment:
This is incredibly complex, and seems to mismatch what I can see covenants being used for in practice.  I anticipate fees being high in future, such that people will do a reasonable amount of engineering to minimize their total footprint.  In particular, they will want to *add fees* after commitment, and want to *batch transactions* using stacking.

The first case implies you want to *exclude* a specific input and output, to allow for fees, or at least allow binding not to cover the final input/output.  The second case implies you want to mul/divide an input number to get the corresponding range of outputs.

The simplest case is a single input and output pair: a-la SIGHASH_SINGLE.  This both allows almost arbitrary fee inputs/outputs, and stacking.

But what if you want to bind a pair of inputs to one output?  Or a pair of outputs to one input?  Both seem reasonably common things to want to do (e.g. opening a dual-funded lightning channel, and closing a channel).

That means you need to be able to select outputs as "current input index / 2" or "current input index * 2 and current input index * 2 + 1".  Numbers other than 2 are possible but this is the most likely case (since, in order to stack, all txs must be of same input-output number form, and I consider 1 and 2 by far the most likely numbers here).

Reply:
Yeah this is true. Initially I didn't have relative indices. I'm still not entirely convinced they are useful. Precisely for the 1-in 2-out case which seems super common to me. I heard "you'd be surprised how easy it is to add an extra input".

My initial thought was that private aggregation (i.e. not through public broadcast media like mempools) would be easily possible as a user can just create/sign a thousand variants of their txs, for each possible input index. This works with absolute indices and doesn't need relative indices. It might even work with public broadcast.

The problem is that doing this with absolute indices only works if everyone in the protocol has the same in-out ratio. (Everyone needs 1-to-2 so you can sign 1in1,2out, 2in3,4out, 3in5,6out, etc). With relative indices, you can sign XinX,1out, XinX,2out, XinX,3out,.. and this way the coordinator can put you in any place and put your second output in some arbitraty place and pick your signatures based on where your second output is placed.

Reply:
This leaves some protocols vulnerable to the partial signature attacks.  Say the covenant requires your outputs go half to pubkeyA and half to pubkeyB.  Now I have two identical 1BTC covenant UTXO inputs, but re-use the same outputs to satisfy both, and steal the other 1BTC.

The same problem applies to "tell me the outputs in the witness data".